src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
+    {
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+        auto *session = BitTorrent::Session::instance();
+        // torrent already exists in transfer list
+        if (const BitTorrent::Torrent *const torrent = session->findTorrent(infoHash); torrent)
+        {
+            // torrent's metadata may not have downloaded yet
+            if (torrent->info().isValid())
+            {
+                setResult(serializeTorrentInfo(*torrent));
+            }
+            else
+            {
+                setStatus(APIStatus::Async);
+                setResult(serializeInfoHash(infoHash));
+            }
+
+            return;
+        }
+
+
+        // metadata may have already been downloaded under a different source URI
+        if (isMetadataDownloaded(infoHash))
+        {
+            // make torrent available for download (via /add API) using this
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+            return;
+        }
+
+        if (!session->isKnownTorrent(infoHash))
+        {
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr)) [[unlikely]]
+            {
+                // our checks above should prevent ever hitting this
+                qDebug("Unable to fetch metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata for '%1'").arg(infoHash.toTorrentID().toString()));
+            }
+
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
+        }
+
+        return;
+    }
+
+    throw APIError(APIErrorType::BadParams, tr("Unable to parse '%1'").arg(source));
+}
+
+void TorrentsController::parseMetadataAction()
+{
+    const DataMap &uploadedTorrents = data();
+    // must provide some value to parse
+    if (uploadedTorrents.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify torrent file(s)"));
+
+    QJsonObject result;
+    for (auto it = uploadedTorrents.constBegin(); it != uploadedTorrents.constEnd(); ++it)
+    {
+        if (const auto loadResult = BitTorrent::TorrentDescriptor::load(it.value()))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+            const QString &fileName = it.key();
+            m_torrentSource.insert(FILE_SOURCE_PREFIX + fileName, infoHash);
```

Comment:
Wouldn't it be easier to just not store such pseudo sources?

Reply:
This is an optimization to prevent the WebUI from having to re-upload the torrent file. When calling `/add`, the caller just has to provide the torrent hash.

Reply:
> This is an optimization to prevent the WebUI from having to re-upload the torrent file.

OK, I see it.


> When calling `/add`, the caller just has to provide the torrent hash.

I would use torrent ID as "source" instead of such "filename".

Reply:
> I would use torrent ID as "source" instead of such "filename".

That's interesting, I like that too. Let me look into that a bit.

Reply:
[Updated](https://github.com/qbittorrent/qBittorrent/compare/84e4a903981d07063645bd0f281dd9854827f493..4a9ea5799bf963726829cb17a6ca8555ca62311c) and it's definitely cleaner.